### PR TITLE
Subscription plan retirement

### PR DESCRIPTION
**Pull request by TChukwuleta**
Adds the ability to retire subscription plans, preventing new subscriptions and migrations to a retired plan while giving admins control over what happens to existing subscribers.

When Plan in retired state:

- No new subscriptions can be created on a retired plan
- No existing subscriber can upgrade or downgrade to a retired plan
-admin can restore a retired plan to active at any time

When retiring a plan with active subscribers, an admin chooses what happens to those subscribers. Options include:

- **Do nothing** subscribers continue renewing as normal
- **Turn off auto-renew** subscribers keep access until their current period ends, then expire
- **Immediately suspend all active subscribers** 
- **Migrate all** move subscribers to another plan, either at period end or immediately

For immediate migration:
- Credit balance is adjusted (unused period refunded, new plan charged)
- If a subscriber can't afford the new plan, they expire and the new plan is queued so "Pay Now" takes them to the correct plan rather than back to the retired one

Other improvements covered in PR

- Retired plans are hidden from the "Add subscriber" modal
- Retired plans don't appear as upgrade/downgrade options in the subscriber portal
- `CreatePlanMigrationCheckout` rejects retired plans
- Plan changes (upgrade/downgrade config) in `AddPlan` already filter to active plans only
- "Create subscriber" dropdown is hidden for retired plans in plan view

Cc. @NicolasDorier

### CODE REVIEW

File: BTCPayServer/Plugins/Subscriptions/SubscriptionHostedService.cs [hunk 1]
```diff
@@ -165,6 +165,62 @@ InvoiceEventCode.MarkedInvalid or
...
+            }
+        }
+        await ctx.SaveChangesAsync();
+        await UpdateSubscriptionStates(subCtx, new MemberSelector.PassedDate(null, now.AddSeconds(1)));
+        foreach (var sub in expiredSubscribers)
+        {
+            sub.NewPlanId = targetPlanId;
+            sub.NewPlan = targetPlan;
+        }
+        await ctx.SaveChangesAsync();
```
[the remark this answers is not in the document]
Comment (TChukwuleta) [reply BTCPayServer/Plugins/Subscriptions/SubscriptionHostedService.cs:222]: I dont want to introduce transaction

But regarding this if the second save fails, the subscriber just expires cleanly without a queued plan. They can still upgrade manually from the portal.

**Comment by NicolasDorier** [comment]
@golyalpha would you tell us about what you think about this approach for plan retirement?

**Comment by golyalpha** [comment]
@NicolasDorier based on the description it looks like about what I'd want to be able to set up in terms of subscription retirements. 

Maybe the only thing I would change would be the ability for an admin to assign an already retired subscription (if I read it correctly, that seems to be explicitly prevented by hiding them as options - IMO they should be differentiated, possibly greyed out, but available to select). If there is still a way to do it via the WebUI some other way as well, then there's no additional change needed.

**Comment by NicolasDorier** [comment]
Ping @Ou7law007 as well, does it seems what you'd expect?

**Comment by Ou7law007** [comment]
This looks very close to what we talked about in https://github.com/btcpayserver/btcpayserver/issues/7023

-----
Is it intended that we're able to migrate from plan_a to plan_a?

image

It does not seem to make sense to allow migration to the same plan. This seems to only happen when there's just 1 plan to choose from that's the current plan being retired.

-----

Believe it or not, this still does not solve the very original issue: https://github.com/btcpayserver/btcpayserver/issues/7145

Maybe it's not intended to do that but the whole idea of 'plan retirement' was to solve this 😅

See https://github.com/btcpayserver/btcpayserver/issues/7145#issuecomment-3814719676

But this is going in the right direction.

-----

Also, the user can still subscribe to a retired plan and force it from being 'Retired' back into 'Retiring' by having an active invoice link already generated before the plan retires.

I don't know if this is intended, but it seems like a bug. I don't know exactly how such cases should be handled where the user pays after a services has been disabled. This is an edge case.

To reproduce:

- Create plan A
- Create plan B
- Allow upgrades and downgrades between A and B
- Create a subscription (subscriber) to A (the cheaper plan)
- Go to portal
- Upgrade to plan B
- Before paying the invoice retire plan B
- Plan B is now retired
- Pay the invoice
- User is allowed to subscribed to a retired plan
- Plan B goes from 'Retired' to 'Retiring'


-----

**Comment by TChukwuleta** [comment]
thank you @golyalpha  for feedback and @Ou7law007 I think that was a bug.. thanks for pointing it out. 

Would clean it up

**Comment by golyalpha** [comment]
Thanks for the rather thorough real testing done here @Ou7law007 

I should've probably been clearer that I haven't looked at the code or tested anything here - my opinion was purely on how the change was described.

**Comment by TChukwuleta** [comment]
Thank you @Ou7law007  good catch
'plan_a to plan_a migration' as well as 'Retired to Retiring via a pre-paid invoice' have been fixed

PR was absolutely based on conversation in https://github.com/btcpayserver/btcpayserver/issues/7023

Looking at https://github.com/btcpayserver/btcpayserver/issues/7145, the core of it based on the conversation is the deeper model issue: subscriber fuses the customer (email, balance) and the subscription (plan binding), Solving that properly means splitting them into separate objects with proration on deletion

This can then be resolved in a follow up PR once @NicolasDorier settles on the approach he mentioned he was weighing

**Comment by NicolasDorier** [comment]
Yeah, refactoring the ability to have subscribers without plan isn't easy at this point. I need to think more about it.

> Maybe the only thing I would change would be the ability for an admin to assign an already retired subscription (if I read it correctly, that seems to be explicitly prevented by hiding them as options - IMO they should be differentiated, possibly greyed out, but available to select). If there is still a way to do it via the WebUI some other way as well, then there's no additional change needed.

That might work without too much disruption to current model. But better to keep that in different PR.

**Comment by Ou7law007** [comment]
Just to be clear, I also did not do any real testing here. I barely ran the PR and checked it out and came up with a few obvious scenarios and edge cases and tried them manually and was able to find these 2 bugs within <5 minutes, so statistically, there's probably more bugs.

Maybe if we can get AI to give us these test scenarios (all possible ones?), then bugs like these may be easier to catch.  I don't know, maybe you can even make AI write these test scenarios, after manually testing them and making sure they work.

People have different approaches to AI, and we're still learning what works and what doesn't, but in my opinion, writing unit tests with AI for things that you know for a fact work (because you tested them manually) allows you to kind of pin the code to something that you know works so that future changes do not require you to manually test things over and over again.

What do you think? @TChukwuleta @NicolasDorier @golyalpha
